### PR TITLE
ITabCompleterBuilder

### PR DESCRIPTION
**Pull request by rkdrns4747**

### CODE REVIEW

**Comment by rkdrns4747** [comment]
Fixing errors.

**Comment by rkdrns4747** [comment]
@wysohn 
Want you to check 4fc87e7 push when reviewing this. I implemented the removed code with builder logic, still cannot guarantee they are in same logic.

**Comment by wysohn** [comment]
@rkdrns4747 
Hey, this looks really good! thank you for the hard work.

I am getting some heavy load from schoolwork, so I will be checking on it as soon as possible.

Just looking at a glance, just as a suggestion, wouldn't it be looking better to use a static method to create a new builder instance than using the `new` keyword?

For example, making the constructor `private`, and create a public static function that merely calls the private constructor which can't be accessed directly outside the scope of the class.

```
(new Builder("abc"))
    .a()
    .b()   
    .builder()
```
or
```
Builder.for("abc")
    .a()
    .b()   
    .builder()
```
 I will leave it up to you

---

Also `ITabCompleterBuilder` sounds a bit awkward since the `I` in front suggests that it's an interface, so how about naming it `TabCompleterBuilder` instead?

**Comment by rkdrns4747** [comment]
@wysohn gotcha.

**Review by wysohn — CHANGES_REQUESTED** [review]
Rofl I just realized that I didn't 'submit' the review

no wonder why you didn't respond :0

Sorry can you take a look at it these? Thanks!

**Comment by wysohn** [comment]
@rkdrns4747 
If I don't get any response by this week, I will push it to the next release

**Comment by rkdrns4747** [comment]
I'll fix it on sunday. Already checked your comment, but there's still no time to sit in front of monitor😔

**Comment by wysohn** [comment]
@rkdrns4747 
Any updates yet?

**Comment by rkdrns4747** [comment]
looks like we have to use it without fixing these things for this moment. Let's push it on development, and fix the issues later when I can finally use keyboard (´ー｀)

**Comment by wysohn** [comment]
> looks like we have to use it without fixing these things for this moment. Let's push it on development, and fix the issues later when I can finally use keyboard (´ー｀)

Sounds good! I will push it to 3.3.x

File: core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java [hunk 1]
```diff
@@ -95,4 +39,129 @@ static ITabCompleter simple(String... arg) {
...
+                this.candidate = new ArrayList<>();
+            }
+            if(template == Template.PLAYER){
+                this.hint = list("<player>");
+                this.candidate = null; // returning null signals to list online players instead
+                //TODO not sure if Sponge does the same
+            }
+        }
+
+        public static TabCompleterBuilder self(){
```
[the remark this answers is not in the document]
Comment (rkdrns4747) [reply core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java:73]: how 'bout instance() ? I have no idea which word I should use.

[the remark this answers is not in the document]
Comment (rkdrns4747) [reply core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java:73]: @wysohn bump.

[the remark this answers is not in the document]
Comment (rkdrns4747) [reply core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java:73]: Or, how about just removing instance() method, and just using Builder.of()? I think this way is more ideal way.

[the remark this answers is not in the document]
Comment (wysohn) [reply core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java:73]: I think that should do it too

Sorry for some reason I didn't see your comment back then.

[the remark this answers is not in the document]
Comment (rkdrns4747) [reply core/src/main/java/io/github/wysohn/triggerreactor/core/manager/trigger/command/ITabCompleter.java:73]: Nah, It's alright, I've been offline like almost 5 month. 😅

**Comment by wysohn** [comment]
Looking good

I will do the detailed review soon, but school is kinda busy

So please bear with me will be back to it soon

**Comment by rkdrns4747** [comment]
> Looking good, thanks a lot!
> 
> I will do the detailed review soon, but school is kinda busy
> 
> So please bear with me will be back to it soon

aight. Then I'll start to work on NameSpace Feature.

**Review by wysohn — APPROVED** [review]
Tested it in-game and seems to be working great. Thank you!

Please go ahead and merge.

**Pull request merged**